keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1190 @@
+- Do not violate version skew, even during node-by-node rollbacks
+
+The control plane:
+
+- Can safely canary components and nodes at the new version for an extended
+  period of time, or to pause an upgrade at any step durning an upgrade. This
+  enhancement, combined with
+  [UVIP](../4020-unknown-version-interoperability-proxy) helps achieve this.
+  
+
+### Non-Goals
+
+- Change the default leader election for components.
+
+## Proposal
+
+- Offer an opt-in leader election mechanism to:
+  - Elect the candidate with the oldest version available.
+  - Provide a way to preempt the current leader.
+  - Reuse the existing lease mechanism as much as possible.
+
+### Component Identity Leases
+
+Components will create identity leases similar to those used by apiserver
+identity, e.g.:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/binary-version: "1.29"
+    coordination.k8s.io/can-lead-leases: kube-system/some-custom-controller
+    coordination.k8s.io/compatibility-version: "1.29"
+  name: some-custom-controller-0001A
+  namespace: kube-system
+spec:
+  holderIdentity: some-custom-controller-0001A
+  leaseDurationSeconds: 10
+  renewTime: "2023-12-05T02:33:08.685777Z"
+```
+
+A component identity lease announces candidacy for leadership by including
+`coordination.k8s.io/can-lead-leases` it is identity lease. If the lease
+expires, the component is considered unavailable for leader election purposes.
+
+### Coordinated Election Controller
+
+A new Coordinated Election Controller will reconcile component Leader Leases
+(primary resource) and Identity Leases (secondary resource, changes trigger
+reconciliation of related leader leases).
+
+Coordinated Election Controller reconciliation loop:
+
+- If no leader lease exists for a components:
+  - Elect leader from candidates by preparing a freshly renewed lease with:
+    - `spec.holderIdentity` set to the identity of the elected leader
+    - `coordination.k8s.io/elected-by: leader-election-controller` (to make
+      lease types easy to disambiguate)
+- If there is a better candidate than current leader:
+  - Sets `coordination.k8s.io/end-of-term: true` on the leader lease, signaling
+    that the leader should stop renewing the lease and yield leadership
+
+```mermaid
+flowchart TD
+   A[Reconcile] --> |Process Leader Lease| B
+   B{Lease Status?} --> |Better Leader Exists| D
+   B --> |Expired/Missing| E
+   D[End Lease Term]
+   E[Elect Leader]
+```
+
+Example of a lease created by Coordinated Election Controller:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/elected-by: coordinated-election-controller
+  name: some-custom-controller
+  namespace: kube-system
+spec:
+  holderIdentity: controller-a
+  leaseDurationSeconds: 10
+  leaseTransitions: 0
+  renewTime: "2023-12-05T18:58:31.295467Z"
+```
+
+The Coordinated Election Controller will run in the kube-apiserver.
```

Comment:
... which turns this mechanism into a full API. Why use annotations so heavily?

Reply:
It allows us to reuse the existing lease mechanism as much as possible. We make use of all Lease fields for component leases, and it seems better to reuse + add annotations rather than introduce a new similar type.

Reply:
Not a new type, but extend it to cover the use-case. Using annotations is cheap, but pretty much an Kubernetes API anti-pattern.

Reply:
ack, we'll get rid of the annotations and extend the Lease API fields instead.